packages/script/src/script-factory.ts

```diff
@@ -0,0 +1,141 @@
+import type { BytesLike } from '@ethersproject/bytes';
+import { Logger } from '@ethersproject/logger';
+import { Interface } from '@fuel-ts/abi-coder';
+import type { JsonAbi, InputValue } from '@fuel-ts/abi-coder';
+import type { BN } from '@fuel-ts/math';
+import type {
+  BuildScriptOptions,
+  CoinQuantityLike,
+  Provider,
+  TransactionResponse,
+  TransactionResult,
+} from '@fuel-ts/providers';
+import { getDecodedLogs, ScriptTransactionRequest } from '@fuel-ts/providers';
+import { MAX_GAS_PER_TX, ReceiptType } from '@fuel-ts/transactions';
+import { versions } from '@fuel-ts/versions';
+import type { BaseWalletLocked } from '@fuel-ts/wallet';
+
+import { Script } from './script';
+
+const logger = new Logger(versions.FUELS);
+
+const FUNCTION_FRAGMENT_NAME = 'main';
+
+type Result<T> = {
+  value: T | BN | undefined;
+  logs: unknown[];
+};
+
+export class ScriptFactory<TInput, TOutput> {
+  bytecode: BytesLike;
+  script: Script<InputValue<void>[], Result<TOutput>>;
+  provider: Provider;
+  interface: Interface;
+  wallet!: BaseWalletLocked | null;
+
+  constructor(bytecode: BytesLike, abi: JsonAbi, walletOrProvider: BaseWalletLocked | Provider) {
+    this.bytecode = bytecode;
+    this.interface = new Interface(abi);
+
+    if (walletOrProvider && 'provider' in walletOrProvider) {
+      this.provider = walletOrProvider.provider;
+      this.wallet = walletOrProvider;
+    } else {
+      this.provider = walletOrProvider;
+      this.wallet = null;
+    }
+
+    this.script = new Script(
+      bytecode,
+      (args: InputValue<void>[]) =>
```

Comment:
Extract this to another function in the class would help with code readability if this is intended to be use only internal we can make it private

Reply:
Maybe create a `buildScript`?